include/circt/Support/InstanceGraphInterface.td

```diff
@@ -0,0 +1,67 @@
+//===- InstanceGraphInterface.td - Interface for instance graphs --------*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file contains interfaces and other utilities for interacting with the
+// generic CIRCT instance graph.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CIRCT_SUPPORT_INSTANCEGRAPH_INSTANCEGRAPHINTERFACE_TD
+#define CIRCT_SUPPORT_INSTANCEGRAPH_INSTANCEGRAPHINTERFACE_TD
+
+include "mlir/IR/AttrTypeBase.td"
+include "mlir/IR/OpBase.td"
+
+def InstanceGraphInstanceOpInterface : OpInterface<"InstanceOpInterface"> {
+  let description = [{
+    This interface provides hooks for an instance-like operation.
+  }];
+  let cppNamespace = "::circt::igraph";
+
+  let methods = [
+    InterfaceMethod<"Get the name of the instance",
+    "::llvm::StringRef", "getInstanceName", (ins)>,
+
+    InterfaceMethod<"Get the name of the instance",
+    "::mlir::StringAttr", "getInstanceNameAttr", (ins)>,
+
+    InterfaceMethod<"Get the name of the instantiated module",
+    "::llvm::StringRef", "getReferencedModuleName", (ins),
+    /*methodBody=*/[{}],
+    /*defaultImplementation=*/[{ return $_op.getModuleName(); }]>,
+
+    InterfaceMethod<"Get the name of the instantiated module",
+    "::mlir::StringAttr", "getReferencedModuleNameAttr", (ins),
+    /*methodBody=*/[{}],
+    /*defaultImplementation=*/[{ return $_op.getModuleNameAttr().getAttr(); }]>,
+
+    InterfaceMethod<"Get the referenced module",
```

Comment:
I don't think we should have this operation without taking a symbol table.  Code which uses this without a symbol table tends to have n^2 behavior.

Reply:
Not using a symbol table also has significant foot-guns related to passes and threading.  I've added a note.